File: src/main/java/com/codingspezis/android/metalonly/player/utils/ImageLoader.java
package com.codingspezis.android.metalonly.player.utils;

import android.content.Context;
import android.graphics.Bitmap;
import android.os.Handler;
import android.widget.ImageView;

import com.codingspezis.android.metalonly.player.R;

import java.io.InputStream;
import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.Collections;
import java.util.Map;
import java.util.WeakHashMap;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

/**
 * this class comes in the original form from:
 * https://github.com/thest1/LazyList
 * <p/>
 * it works not longer with URLs but with getModerator names of metal only some
 * synchronized statements are also added
 */
public class ImageLoader {

    final int stub_id = R.drawable.mo_wait;
    private final Map<ImageView, String> imageViews =
            Collections.synchronizedMap(new WeakHashMap<ImageView, String>());

    private final MemoryCache memoryCache = new MemoryCache();
    private final FileCache fileCache;
    private final ExecutorService executorService;
    private final Handler handler = new Handler();

    public ImageLoader(Context context) {
        fileCache = new FileCache(context);
        executorService = Executors.newFixedThreadPool(5);
    }

    public void displayImage(String moderator, ImageView imageView) {
        imageViews.put(imageView, moderator);

        Bitmap bitmap = memoryCache.get(moderator);

        if (bitmap != null) {
            imageView.setImageBitmap(bitmap);
        } else {
            queuePhoto(moderator, imageView);
            imageView.setImageResource(stub_id);
        }
    }

    private void queuePhoto(String moderator, ImageView imageView) {
        ModImageLoadingQueueItem p = new ModImageLoadingQueueItem(moderator, imageView);
        executorService.submit(new PhotosLoader(p));
    }

    synchronized Bitmap getBitmap(String moderator) {
        Bitmap b = fileCache.decodeMod(moderator);
        if (b != null) {
            return b;
        }

        try {
            URL imageUrl = new URL(UrlConstants.METAL_ONLY_MODERATOR_PIC_BASE_URL + moderator);

            HttpURLConnection conn = (HttpURLConnection) imageUrl.openConnection();
            conn.setConnectTimeout(30000);
            conn.setReadTimeout(30000);
            conn.setInstanceFollowRedirects(true);

            InputStream is = conn.getInputStream();
            OutputStream os = fileCache.getOutputStream(moderator);

            Utils.CopyStream(is, os);

            is.close();
            os.close();

            return fileCache.decodeMod(moderator);
        } catch (Throwable ex) {
            ex.printStackTrace();
            if (ex instanceof OutOfMemoryError) {
                memoryCache.clear();
            }
            return null;
        }
    }

    boolean imageViewReused(ModImageLoadingQueueItem modImageLoadingQueueItem) {
        String tag = imageViews.get(modImageLoadingQueueItem.getImageView());
        return tag == null || !tag.equals(modImageLoadingQueueItem.getModerator());
    }

    class PhotosLoader implements Runnable {
        ModImageLoadingQueueItem modImageLoadingQueueItem;

        PhotosLoader(ModImageLoadingQueueItem modImageLoadingQueueItem) {
            this.modImageLoadingQueueItem = modImageLoadingQueueItem;
        }

        @Override
        public void run() {
            try {
                if (imageViewReused(modImageLoadingQueueItem)) {
                    return;
                }
                Bitmap bmp = getBitmap(modImageLoadingQueueItem.getModerator());
                memoryCache.put(modImageLoadingQueueItem.getModerator(), bmp);
                if (imageViewReused(modImageLoadingQueueItem)) {
                    return;
                }
                BitmapDisplayer bd = new BitmapDisplayer(ImageLoader.this, bmp,
                        modImageLoadingQueueItem);
                handler.post(bd);
            } catch (Throwable th) {
                th.printStackTrace();
            }
        }
    }

}

File: app/src/main/kotlin/com/codingspezis/android/metalonly/player/utils/ImageLoader.kt
package com.codingspezis.android.metalonly.player.utils

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.os.Handler
import android.widget.ImageView
import com.codingspezis.android.metalonly.player.R
import java.net.HttpURLConnection
import java.net.URL
import java.util.Collections
import java.util.WeakHashMap
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

/**
 * this class comes in the original form from:
 * https://github.com/thest1/LazyList
 *
 *
 * it works not longer with URLs but with getModerator names of metal only some
 * synchronized statements are also added
 */
class ImageLoader private constructor(context: Context) {

    private val imageViews = Collections.synchronizedMap(WeakHashMap<ImageView, String>())

    private val memoryCache: Cache = MemoryCache()
    private val fileCache: Cache = FileCache(context)

    /** @todo use actual number of cpu cores here */
    private val executorService: ExecutorService = Executors.newFixedThreadPool(8)
    private val handler = Handler()

    fun loadImage(moderator: String, imageView: ImageView) {
        val queueItem = ModImageLoadingQueueItem(moderator, imageView)
        if (imageViewReused(queueItem))
            return

        imageViews.put(imageView, moderator)

        val memCachedBm = memoryCache[moderator]
        if (memCachedBm != null) {
            replaceImage(imageView, memCachedBm)
            return
        }

        val fileCachedBm = fileCache[moderator]
        if (fileCachedBm != null) {
            memoryCache[moderator] = fileCachedBm
            replaceImage(imageView, fileCachedBm)
            return
        }

        executorService.submit(ImageDownloader(queueItem))
    }

    internal fun replaceImage(imageView: ImageView, bitmap: Bitmap) {
        imageView.post({
            imageView.setImageBitmap(bitmap)
        })
    }

    internal fun imageViewReused(modImageLoadingQueueItem: ModImageLoadingQueueItem): Boolean {
        val tag = imageViews[modImageLoadingQueueItem.imageView]
        return tag == modImageLoadingQueueItem.moderator
    }

    internal inner class ImageDownloader(var queueItem: ModImageLoadingQueueItem) : Runnable {

        override fun run() {
            try {
                val bitmap = downloadImage(queueItem.moderator)

                memoryCache[queueItem.moderator] = bitmap

                val imageView = queueItem.imageView
                imageView.post {
                    if (bitmap != null) {
                        imageView.setImageBitmap(bitmap)
                    } else {
                        imageView.setImageResource(R.drawable.image_bg)
                    }
                }

                val bd = BitmapDisplayer(this@ImageLoader, bitmap, queueItem)
                handler.post(bd)
            } catch (th: Throwable) {
                th.printStackTrace()
            }
        }

        @Synchronized internal fun downloadImage(moderator: String): Bitmap? {
            try {
                val imageUrl = URL(UrlConstants.METAL_ONLY_MODERATOR_PIC_BASE_URL + moderator)

                val conn = imageUrl.openConnection() as HttpURLConnection
                conn.connectTimeout = 30000
                conn.readTimeout = 30000
                conn.instanceFollowRedirects = true

                val inputStream = conn.inputStream

                val bitmap = BitmapFactory.decodeStream(inputStream, null, BitmapFactory.Options())
                fileCache[moderator] = bitmap
                inputStream.close()

                return fileCache[moderator]
            } catch (ex: Throwable) {
                ex.printStackTrace()
                if (ex is OutOfMemoryError) {
                    memoryCache.clear()
                }
                return null
            }
        }
    }

    companion object {

        private val lock: Object = Object()

        private var INSTANCE: ImageLoader? = null

        /**
         * Gets a singleton instance of the ImageLoader for usage.
         */
        fun instance(context: Context) : ImageLoader {
            synchronized(lock) {
                if (INSTANCE == null) {
                    INSTANCE = ImageLoader(context.applicationContext)
                }
            }
            return INSTANCE!!
        }
    }
}
